simulation/chart.py

```diff
@@ -1,232 +1,438 @@
+        foreground: Optional[Color] = None,
+        background: Optional[Color] = None,
+        border: Optional[Union[bool, Tuple[int, int, int]]] = None,
+    ) -> None:
+        self.__save = output_file is not None
+
+        # Set up PyQtGraph
+        if foreground is not None:
+            pg.setConfigOption("foreground", _make_color(foreground))
+        if background is not None:
+            pg.setConfigOption("background", _make_color(background))
+        if antialias is not None:
+            pg.setConfigOptions(antialias=antialias)
+
+        self.__app = pg.mkQApp(title)
+
+        self.__layout = pg.GraphicsLayoutWidget(
+            show=not self.__save, title=title, size=size, border=border
+        )
 
-    # May have to rescale (HiDPI displays, etc)
-    if (qimage.width(), qimage.height()) != size:
-        qimage = (
-            QtGui.QPixmap.fromImage(qimage)  # type: ignore
-            .scaled(
-                size[0],
-                size[1],
-                mode=QtCore.Qt.TransformationMode.SmoothTransformation,  # type: ignore
+        self.__init_size = size
+
+        if output_file is not None:
+            self.__ffmpeg_process: Popen = (
+                ffmpeg.input(
+                    "pipe:",
+                    format="rawvideo",
+                    pix_fmt="rgb24",
+                    s=f"{size[0]}x{size[1]}",
+                )
+                .output(output_file, vcodec=output_codec, pix_fmt=output_pixel_format)
+                .overwrite_output()
+                .run_async(pipe_stdin=True)
             )
-            .toImage()
-            .convertToFormat(
+
+    def create_chart(
+        self,
+        title: Optional[str] = None,
+        height: Optional[int] = None,
+        legend_width: Optional[int] = None,
+        x_label: Optional[str] = None,
+        x_range: Optional[Tuple[float, float]] = None,
+        x_log: Optional[bool] = None,
+        y_label: Optional[str] = None,
+        y_range: Optional[Tuple[float, float]] = None,
+        y_log: Optional[bool] = None,
+    ) -> ChartWidget:
+        """Add a chart with legend for plots as a row
+
+        Args:
+            title (str, optional): Title to display on the top. Defaults to None.
+            height (int, optional): Prefred height. Defaults to None.
+            legend_width (int, optional):Legend width. Defaults to None.
+            x_label (str, optional): Label text to display under x axis. Defaults to None.
+            x_range (Tuple[float,float], optional): Constant range(min,max value) of x axis. Defaults to None.
+            x_log (bool, optional): Use logarithmic scale to display x axis. Defaults to None.
+            y_label (str, optional): Label text to display next to y axis. Defaults to None.
+            y_range (Tuple[float,float], optional): Constant range(min,max value) of y axis. Defaults to None.
+            y_log (bool, optional): Use logarithmic scale for y axis. Defaults to None.
```

Comment:
Typically, you want a newline between the `Args` and the `Returns`.

Reply:
https://github.com/semiotic-ai/autoagora-agents/commit/543ffd999da9d9e71b97c69fd84b59c0dec6a395